keps/sig-network/0007-ipvs-proxier.md

```diff
@@ -0,0 +1,571 @@
+---
+kep-number: 7
+title: IPVS Load Balancing Mode in Kubernetes
+status: implemented
+authors:
+    - "@rramkumar1"
+owning-sig: sig-network
+reviewers:
+  - "@thockin"
+  - "@m1093782566"
+approvers:
+  - "@thockin"
+  - "@m1093782566"
+editor:
+  - "@thockin"
+  - "@m1093782566"
+creation-date: 2018-03-21
+---
+
+# IPVS Load Balancing Mode in Kubernetes
+
+**Note: This is a retroactive KEP. Credit goes to @m1093782566, @haibinxie, and @quinton-hoole for all information & design in this KEP.**
+
+**Important References: https://github.com/kubernetes/community/pull/692/files**
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non\-goals](#non-goals)
+* [Proposal](#proposal)
+  * [Kube-Proxy Parameter Changes](#kube-proxy-parameter-changes)
+  * [Build Changes](#build-changes)
+  * [Deployment Changes](#deployment-changes)
+  * [Design Considerations](#design-considerations)
+    * [IPVS service network topology](#ipvs-service-network-topology)
+    * [Port remapping](#port-remapping)
+    * [Falling back to iptables](#falling-back-to-iptables)
+    * [Supporting NodePort service](#supporting-nodeport-service)
+    * [Supporting ClusterIP service](#supporting-clusterip-service)
+    * [Supporting LoadBalancer service](#supporting-loadbalancer-service)
+    * [Session Affinity](#session-affinity)
+    * [Cleaning up inactive rules](#cleaning-up-inactive-rules)
+    * [Sync loop psuedo code](#sync-loop-psuedo-code)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks](#drawbacks)
+* [Alternatives](#alternatives)
+
+## Summary
+
+We are building a new implementation of kube proxy built on top of IPVS (IP Virtual Server).
+
+## Motivation
+
+As Kubernetes grows in usage, the scalability of its resources becomes more and more
+important. In particular, the scalability of services is paramount to the adoption of Kubernetes
+by developers/companies running large workloads. Kube Proxy, the building block of service routing
+has relied on the battle-hardened iptables to implement the core supported service types such as
+ClusterIP and NodePort. However, iptables struggles to scale to tens of thousands of services because
+it is designed purely for firewalling purposes and is based on in-kernel rule chains. On the
+other hand, IPVS is specifically designed for load balancing and uses more efficient data structures
+under the hood. For more information on the performance benefits of IPVS vs. iptables, take a look
+at these [slides](https://docs.google.com/presentation/d/1BaIAywY2qqeHtyGZtlyAp89JIZs59MZLKcFLxKE6LyM/edit?usp=sharing).
+
+### Goals
+
+* Improve the performance of services
```

Comment:
Probably worth adding a little bit more detail here. Can we point to or write down concrete performance targets, or is it simply "anything better than what we've got"?

Reply:
The slide deck linked in the intro should have those performance improvements.